docs/design/2021-04-21-unify-log-library.md

```diff
@@ -0,0 +1,141 @@
+# Proposal: Unifying Log Library
+
+- Author(s): [Yifan Xu](https://github.com/SabaPing)
+- Last updated:  Apr 12, 2021
+- Discussion at: https://github.com/pingcap/tidb/pull/24181
+- Tracking issue: https://github.com/pingcap/tidb/issues/24190
+
+## Motivation or Background
+
+Except for slow query logs, all other logs must satisfy the [unified-log-format RFC standard](https://github.com/tikv/rfcs/blob/master/text/2018-12-19-unified-log-format.md).
+
+However, in practice, it was found that the format of logs is confusing:
+
+- There are few logging configuration instructions in the document. We need to enrich document especially the type and the format of logs each component would emit.
+- The configured logging parameters do not match the runtime logging, e.g. `tidb_stderr` is configured with text format, but the log is in json format.
+- The logs of some components do not meet the [unified-log-format RFC standard](https://github.com/tikv/rfcs/blob/23d4f9aed68a295b678e8bd909ee8479e3ba0bd1/text/2018-12-19-unified-log-format.md), e.g. `tiflash_cluster_manager`.
+- Duplicate logs, e.g. `pd_stderr` will emit both text and json logs with duplicate content (but with a few subtle differences in timestamps).
+
+## Logging code for each component
+
+### [pingcap/log](https://github.com/pingcap/log)
+
+As a common logging library in PingCAP, it does the following things:
+
+- Provides the standard config schema.
+- Provides a factory method for creating log handlers.
+- Hard code the log format according to [unified-log-format RFC standard](https://github.com/tikv/rfcs/blob/23d4f9aed68a295b678e8bd909ee8479e3ba0bd1/text/2018-12-19-unified-log-format.md).
+- Encapsulates the logic of the rolling file.
+- Provides global log handler and related methods for package dimension.
+
+### TiDB
+
+Log library dependencies:
+
+![tidb-log-dependency](./imgs/tidb-log-dependency.png)
+
+For historical reasons, TiDB has two third-party logging libraries, `logrus` and `pingcap/log`. `pingcap/log` is a wrapper of `zap`.
+
+Logs of TiDB can be divided into two types, slow query logs and the other logs. 
+As mentioned above, these two types of logs are emitted through two different logging libraries, which results in separate configurations for the two types of logs and requires writing additional configuration conversion code.
```

Comment:
These two logs are intentionally to have different formats. Will there be a problem when merging the two separated configuration into one?

Reply:
I added extra conversion layers. After testing and verifying, it should be fine.